### PR TITLE
feat(catalog): add support for entity policies in the new backend.

### LINKED ISSUE #15455: 🚀 Feature: `catalogProcessingExtensionPoint` supports `addEntityPolicy`
### 🔖 Feature description

Within the [New Backend System](https://backstage.io/docs/plugins/new-backend-system), the new concept of [Extension Points](https://backstage.io/docs/plugins/new-backend-system#extension-points) are introduced and called out to be used when defining plugin [Modules](https://backstage.io/docs/plugins/new-backend-system#modules)

As of the latest release (currently v1.9.1), the `catalogProcessingExtensionPoint` that allows modules to hook into the catalog processing setup [only allows for adding new Processors and Entity Providers](https://github.com/backstage/backstage/blob/v1.9.1/plugins/catalog-node/src/extensions.ts#L23-L30).

The [CatalogBuilder itself supports addition of Entity Policies](https://github.com/backstage/backstage/blob/v1.9.1/plugins/catalog-backend/src/service/CatalogBuilder.ts#L183-L199) and as a plugin module developer, it would be nice to be able to add Entity Policies from backend catalog module plugins

### 🎤 Context

Entity Policies are beneficial and a [key concept of the Entity Processing pipeline](https://backstage.io/docs/features/software-catalog/life-of-an-entity#key-concepts:~:text=Policies%2C%20that%20establish%20baseline%20rules%20about%20the%20shape%20of%20entities%2C) within the life of an entity.

Adding Entity policies is the only [key concept in the life of an entity](https://backstage.io/docs/features/software-catalog/life-of-an-entity#key-concepts) that is missing from the current [catalogProcessingExtensionPoint](https://github.com/backstage/backstage/blob/v1.9.1/plugins/catalog-node/src/extensions.ts#L23-L30).

As such, it feels prudent to include it for module development.

### ✌️ Possible Implementation

Add `addEntityPolicy` to the `CatalogProcessingExtensionPoint`
```diff
# plugins/catalog-node/src/extensions.ts

export interface CatalogProcessingExtensionPoint {
  addProcessor(
    ...processors: Array<CatalogProcessor | Array<CatalogProcessor>>
  ): void;
  addEntityProvider(
    ...providers: Array<EntityProvider | Array<EntityProvider>>
  ): void;
+ addEntityPolicy(
+   ...policies: Array<EntityPolicy | Array<EntityPolicy>>
+ ): void;
}
```

Then add the addition of added policies from the extension point during backend plugin initialization
```diff
# plugins/catalog-backend/src/service/CatalogPlugin.ts

class CatalogExtensionPointImpl implements CatalogProcessingExtensionPoint {
  #processors = new Array<CatalogProcessor>();
  #entityProviders = new Array<EntityProvider>();
+ #entityPolicies = new Array<EntityPolicy>();

  addProcessor(
    ...processors: Array<CatalogProcessor | Array<CatalogProcessor>>
  ): void {
    this.#processors.push(...processors.flat());
  }

  addEntityProvider(
    ...providers: Array<EntityProvider | Array<EntityProvider>>
  ): void {
    this.#entityProviders.push(...providers.flat());
  }

+ addEntityPolicy(
+   ...policies: Array<EntityPolicy | Array<EntityPolicy>>
+ ): void {
+   this.#entityPolicies.push(...policies.flat());
+ }

  get processors() {
    return this.#processors;
  }

  get entityProviders() {
    return this.#entityProviders;
  }

+ get entityPolicies() {
+   return this.#entityPolicies;
+ }
}

...

# in `export const catalogPlugin = createBackendPlugin({`

 const winstonLogger = loggerToWinstonLogger(logger);
 const builder = await CatalogBuilder.create({
   config,
   reader,
   permissions,
   database,
   logger: winstonLogger,
 });
 builder.addProcessor(...processingExtensions.processors);
 builder.addEntityProvider(...processingExtensions.entityProviders);
+builder.addEntityPolicy(...processingExtensions.entityPolicies);
 const { processingEngine, router } = await builder.build();

 await processingEngine.start();
 lifecycle.addShutdownHook({
   fn: async () => {
     await processingEngine.stop();
   },
 });
 httpRouter.use(router);
```

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

closes #15455, pretty straightforward, I think it's worth keeping it with the rest of the `builder.add...` properties for now.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Rugvip — COMMENTED** [review]
Wanna highlight the proposal in https://github.com/backstage/backstage/issues/15455#issuecomment-1453279438, to keep this in a separate `CatalogModelExtensionPoint` (name TBD). It's perhaps slightly different and something to evolve separately from the way processors and providers are installed? In particular also because I think more methods might be needed than just adding new policies.

**Comment by sennyeya** [comment]
@Rugvip Updated to a separate `CatalogModelExtensionPoint` with `addPolicies` and `replacePolicies`. Are there any other chunks we may want to include in this work?

**Comment by freben** [comment]
I have a bit of a discussion point. What should we aim for as the long term goal for this?

Context: One of the key emerging things in the new backend system is that we are trying to start backends _as empty as is possible_. Ie having as few defaults as possible, and making it an intentional thing to explicitly add things to it. Hence one can add a new `addDefaults?: boolean` parameter to an old `createRouter` for example, so that the new backend plugin code can ensure it gets created stripped clean.

I think the policies generally fall in these categories now:

1. Actually important ones
   - `SchemaValidEntityPolicy` ensures that the entity is even remotely the right shape. But interestingly, this is already done by the catalog backend [far, far earlier](https://github.com/backstage/backstage/blob/72341bc0832a12f69a50fa6f90647bd6661be6a2/plugins/catalog-backend/src/processing/DefaultCatalogProcessingOrchestrator.ts#L132) so the check is sort of moot. Well, unless a processor completely borks the entity shape in its pre step, but that gets caught anyway by the validation step later.
   - `DefaultNamespaceEntityPolicy` makes sure that `metadata.namespace` always gets the value `default` if it's unset. This should honestly just be a hard coded thing in the catalog backend anyway.
   - `FieldFormatEntityPolicy` is the recipient for the field format validators, which is an odd indirection. To me it sounds like the new `CatalogModelExtensionPoint` (which I think is a great direction!) should have methods for directly setting field format rules instead of via replacing the policy.
1. Some very dubious ones
   - `GroupDefaultParentEntityPolicy` isn't even used
   - `NoForeignRootFieldsEntityPolicy` well yeah sure ... but again that should be a dedicated method on the `CatalogModelExtensionPoint` then
1. User defined ones. I think we should outright not allow this use case. Possibly we could just allow `add` and immediately marking it as deprecated and telling people to please not do that, or we could offer a policy-to-processor converter function that they can call as a compat feature while they migrate, so they can add them as processors instead.

As mentioned, I still think this particular extension point with this exact name is an awesome thing. If it has some specific great methods for doing model adjustments that are better than "just figure out how to write an entire processor", I think this can be a winner. But - what to do about policies in particular?

I vote for making a policy-to-processor converter function. Then we move toward stripping down the set of policies that are installed by default to an absolute minimum in the new backend system, and figure out how to move off of them entirely. What do you think?

**Comment by sennyeya** [comment]
@freben Appreciate the in depth comment! Definitely agree, I've updated this PR to make `addPolicies` deprecated on release, added a new `policyToProcessorTransformer`, and added support for field validators on the extension point. I think the combination of `setFieldValidators` and `addPolicies` will cover the existing use cases for policies (minus removing one of the default policies, but I think that fits into "don't do this" territory).

File: plugins/catalog-backend/api-report.md
Comment (freben) [inline plugins/catalog-backend/api-report.md:434]: small nit i guess - a more active wording like `transformLegacyPolicyToProcessor` or something might read better

**Comment by sennyeya** [comment]
@freben True, calling it out more clearly in the release notes/as a guide instead of introducing it as deprecated feels more intuitive. There's no current support and all we're saying it that there will be no support going forward as well.

**Comment by freben** [comment]
@sennyeya needs a rebase now it seems

**Comment by freben** [comment]
Oh and it also needs a changeset for -node still

**Comment by freben** [comment]
interestingly, now i have to approve workflow runs what with the new username and all

**Comment by aramissennyeydd** [comment]
@freben Had the same issue with #23036, I'm hoping this is just because it's the first few commits to the project on this account.

**Review by freben — APPROVED** [review]
Nice! Got a verbal ok from catalog maintainers for this too, let's 🚢

**Pull request merged**
